src/server/ua_file.c

```diff
@@ -0,0 +1,419 @@
+    }
+    UA_FileInfo *fileInfo = getFileHandleInfo(server, fileObject, fileHandle);
+    if(!fileInfo)
+    {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+    if(fileInfo->openFileMode != UA_OPENFILEMODE_READ)
+    {
+        return UA_STATUSCODE_BADINVALIDSTATE;
+    }
+    UA_ByteString *readBuffer = UA_ByteString_new();
+    UA_ByteString_init(readBuffer);
+    if(!UA_file_eof(fileInfo->file))
+    {
+        if(requestedLengthToRead < fileObject->fileSize)
+        {
+            lengthToRead = requestedLengthToRead;
+        }
+        else
+        {
+            lengthToRead = fileObject->fileSize;
+        }
+        readBuffer->length = (size_t)lengthToRead;
+        readBuffer->data = (UA_Byte*) UA_malloc(readBuffer->length * sizeof(UA_Byte));
+        UA_file_read(readBuffer->data, (readBuffer->length + 1), 1, fileInfo->file);
+    }
+    UA_Variant_setScalarCopy(output, readBuffer, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    UA_ByteString_delete(readBuffer);
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+UA_Server_writeFileCallback(UA_Server *server,
+                            const UA_NodeId *sessionId, void *sessionHandle,
+                            const UA_NodeId *methodId, void *methodContext,
+                            const UA_NodeId *objectId, void *objectContext,
+                            size_t inputSize, const UA_Variant *input,
+                            size_t outputSize, UA_Variant *output)
+{
+    UA_UInt32 fileHandle = *(UA_UInt32 *)input[0].data;
+    UA_ByteString *fileContentToWrite = (UA_ByteString*)input[1].data;
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_FileType *fileObject = getFileTypeObject(server, objectId);
+    if(!fileObject)
+    {
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+    UA_FileInfo *fileInfo = getFileHandleInfo(server, fileObject, fileHandle);
+    if(!fileInfo)
+    {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+    if(((fileInfo->openFileMode != UA_OPENFILEMODE_ERASEEXISTING) ||
+       (fileInfo->openFileMode != (UA_OPENFILEMODE_ERASEEXISTING | UA_OPENFILEMODE_WRITE))) &&
+       (fileInfo->openFileMode != UA_OPENFILEMODE_READ))
+    {
+        for(size_t i = 0; i < fileContentToWrite->length; i++)
+        {
+            UA_file_print(fileInfo->file, "%c", fileContentToWrite->data[i]);
+        }
+        return retval;
+    }
+    retval = UA_STATUSCODE_BADINVALIDSTATE;
+    return retval;
+}
+
+static UA_StatusCode
+UA_Server_setPositionFileCallback(UA_Server *server,
+                                    const UA_NodeId *sessionId, void *sessionHandle,
+                                    const UA_NodeId *methodId, void *methodContext,
+                                    const UA_NodeId *objectId, void *objectContext,
+                                    size_t inputSize, const UA_Variant *input,
+                                    size_t outputSize, UA_Variant *output)
+{
+    return UA_STATUSCODE_BADNOTIMPLEMENTED;
```

Comment:
The idea of a file is to transport content that is larger than what fits inside a single message.
You can only do that if you preserve the position in the file across different calls to `read`.

So you need some persistency of that information stored in the user session.
There is a way to get there in a clean fashion. The new PR #4463 adds a key-value configuration system.
Add key-value parameters to the session struct.
And select a nice key-name under which the position for each particular file is stored.

Reply:
Also limit the number of open files per session.
Or this can be a way to consume arbitrary amounts of memory.

Reply:
what should be the (good) default value of max open files per session ?

Reply:
> The idea of a file is to transport content that is larger than what fits inside a single message. You can only do that if you preserve the position in the file across different calls to `read`.
> 
> So you need some persistency of that information stored in the user session. There is a way to get there in a clean fashion. The new PR #4463 adds a key-value configuration system. Add key-value parameters to the session struct. And select a nice key-name under which the position for each particular file is stored.

Is the persistency of the current position of file descriptor in the session is really needed here ? Because this seems to be implied by the current implementation anyway. As long the Close is not called, each subsequent call to Read will read from the underlying FIL E* stream starting from it's current stream position.